### PR TITLE
PS2 Keyboard

### PR DESCRIPTION
**Pull request by Sploder12**
# Description

Implements the PS2 interface along with scan code conversion for scan code set 1.

Issue #7 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

### Testing 
I pressed the buttons and the ascii showed up on screen.

**Test Configuration**:
* qemu 7.2.0 on Windows 10 with WSL 5.10.16.3 running Ubuntu 20.04
* python 3.10.10

# Checklist:
	
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by Sploder12** [comment]
Documentation coming soon

File: src/lib/device/keyboard.c [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+    [Key_n] = 'n',          [Key_m] = 'm',          [Key_comma] = ',',
+    [Key_period] = '.',     [Key_slash] = '/',      [Key_space] = ' ',
+    [Key_esc] = 27,         [Key_numDiv] = '/',     [Key_numMult] = '*',
+    [Key_numSub] = '-',     [Key_numAdd] = '+',     [Key_num7] = '7',
+    [Key_num8] = '8',       [Key_num9] = '9',       [Key_num4] = '4',
+    [Key_num5] = '5',       [Key_num6] = '6',       [Key_num1] = '1',
+    [Key_num2] = '2',       [Key_num3] = '3',       [Key_numEnter] = '\n',
+    [Key_num0] = '0',       [Key_numDecimal] = '.'};
+
+// also note the correct usage of what Dr. Blanton tells you not to do.
```
Comment (bicknrown) [inline src/lib/device/keyboard.c:93]: Correct, and here's some history for you if you haven't seen this:
> The rule, which survives in today's C, is that values of array type are converted, when they appear in expressions, into pointers to the first of the objects making up the array.  

https://www.bell-labs.com/usr/dmr/www/chist.html
In the "Embryonic C" section.

Comment (bicknrown) [reply src/lib/device/keyboard.c:93]: as a followup, the reason that we (220 course staff in general) teach that practice is to remove the confusion of attempting to figure out where that line is, while students are already trying to learn the rest of the language.

File: src/lib/device/keyboard.c [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        state->extended = 1;
+        return nonePress;
+    }
+
+    if (code == 0xE1) {
+        state->extended = 3;
+        return nonePress;
+    }
+
+    bool released = code & 0b10000000; // highest bit is set on release
```
Comment (bicknrown) [inline src/lib/device/keyboard.c:227]: i'm on the fence whether or not the masks should also be macros, the comments help, but i am up for discussion on this.

File: src/lib/device/keyboard.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    KeyPress out = {get_ticks(), Key_none, KeyReleased, 0};
+    if (!released)
+        out.event = KeyPressed;
+
+    bool numLock = state->keystates[(unsigned)(Key_numLock)] == KeyToggled;
+
+    uint8_t adjcode = code & 0b01111111; // code without release
```
Comment (bicknrown) [inline src/lib/device/keyboard.c:234]: see previous comment.

File: src/lib/device/ps2.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
+#include "ps2.h"
+
+#include "../../os/hard/idt.h"
+#include "../../os/hard/port_io.h"
+#include "container/ring_buffer.h"
+
+#define PS2_BUF_SIZE 64
```
Comment (bicknrown) [inline src/lib/device/ps2.c:7]: perhaps consolidate macros to the header?

Comment (bicknrown) [reply src/lib/device/ps2.c:7]: that and also the typedefs and structures, unless there is a particular reason you want it this way.

File: src/lib/device/ps2.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    case JPNPKey:
+    case JPNAKey:
+    case SunKey:
+        return (enum DeviceType)(b);
+    default:
+        return Ps2Unknown;
+    }
+}
+
+uint8_t readStat(void) { return inb(PS2_STAT_CMD); }
```
Comment (bicknrown) [inline src/lib/device/ps2.c:129]: another candidate for consistent formatting with other functions.

File: src/lib/device/ps2.c [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+    return readData();
+}
+
+void setConf(uint8_t newconf) {
+    sendCMD(0x60);
+    sendData(newconf);
+}
+
+// public alias of sendData
+bool sendPort1(uint8_t b) { return sendData(b); }
```
Comment (bicknrown) [inline src/lib/device/ps2.c:194]: this also.

File: src/os/main.c [hunk 7]
```diff
@@ -21,7 +24,7 @@ int os_main() {
     static const char test_str[] = "test";
     size_t test_idx = 0;
 
-    while (1 + 2 == 3) {
+    while (1000 == 1000 - 1000 + 1000) {
```
Comment (bicknrown) [inline src/os/main.c:27]: always something :facepalm:

**Review by bicknrown — DISMISSED** [review]
Overall, everything but the changes I marked seems alright to me, thank you for the contributions!

**Comment by bicknrown** [comment]
this branch is also out of date with `main`. between the two PRs, we can agree on an order to service them in, whichever is easier.

File: src/lib/device/ps2.c (hunk 5 above)
Comment (Sploder12) [reply src/lib/device/ps2.c:129]: Issue with the formatter. I'll look into it

File: src/lib/device/ps2.c (hunk 4 above)
Comment (Sploder12) [reply src/lib/device/ps2.c:7]: The typedef I'd like to keep in the .c, It wouldn't be useful outside and it's just an implementation detail.
The macros could be moved, but I feel they are implementation details and wouldn't be very useful.

File: src/lib/device/keyboard.c (hunk 2 above)
Comment (Sploder12) [reply src/lib/device/keyboard.c:227]: My original reason for not making it a macro is that this only applies to Scan Code set 1, and macros have global visibility. But I could also just name it something SC1 specific like `SC1_RELEASE_MASK`. What are your thoughts on doing that?

**Review by Dylanz5 — APPROVED** [review]

**Review by emilkovacev — APPROVED** [review]
I made a few comments, but they're small and not essential. Excellent PR!

**Pull request merged**